Name: Haris Sumra

Assignment: Sparsity

Here, you’re asked to determine the sparsity of four different recommender system datasets: 
1. Jokester.  https://goldberg.berkeley.edu/jester-data/.   
2. MovieLens.  https://grouplens.org/datasets/movielens/.  
3. Your Survey dataset (your week 1 assignment) 
4. A fourth recommender system dataset that you find of your choosing.

### Jokester 

In [94]:
#Import Dependencies
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import os
import tarfile

In [95]:
#Loading the Data from Github Repository
username = "harisx91"
path = "Recommender_System_DAV_6300/master/Module_1/jester-data-1.csv"
df = pd.read_csv(f'https://raw.githubusercontent.com/{username}/{path}')

In [96]:
df.head()

74  -7.82   8.79  -9.66  -8.16  -7.52  -8.5  -9.85  4.17  -8.98  ...  \
0  100   4.08  -0.29   6.36   4.37  -2.38 -9.66  -0.73 -5.34   8.88  ...   
1   49  99.00  99.00  99.00  99.00   9.03  9.27   9.03  9.27  99.00  ...   
2   48  99.00   8.35  99.00  99.00   1.80  8.16  -2.82  6.21  99.00  ...   
3   91   8.50   4.61  -4.17  -5.39   1.36  1.60   7.04  4.61  -0.44  ...   
4  100  -6.17  -3.54   0.44  -8.50  -7.09 -4.32  -8.69 -0.87  -6.65  ...   

   2.82.2  99.18  99.19  99.20  99.21  99.22  -5.63  99.23  99.24  99.25  
0    2.82  -4.95  -0.29   7.86  -0.19  -2.14   3.06   0.34  -4.32   1.07  
1   99.00  99.00  99.00   9.08  99.00  99.00  99.00  99.00  99.00  99.00  
2   99.00  99.00  99.00   0.53  99.00  99.00  99.00  99.00  99.00  99.00  
3    5.19   5.58   4.27   5.19   5.73   1.55   3.11   6.55   1.80   1.60  
4   -3.54  -6.89  -0.68  -2.96  -2.18  -3.35   0.05  -9.08  -5.05  -3.45  

[5 rows x 101 columns]

In [97]:
#Looking for null values
df.isnull().sum()

74       0
-7.82    0
8.79     0
-9.66    0
-8.16    0
        ..
99.22    0
-5.63    0
99.23    0
99.24    0
99.25    0
Length: 101, dtype: int64

In [98]:
print(df.shape)

(24982, 101)


In [99]:
##The values "99.0" corresponds to "null"= "not rated". Replaces those values to 0
df_01= df.replace(99.0, 0)
df_01.head()

74  -7.82  8.79  -9.66  -8.16  -7.52  -8.5  -9.85  4.17  -8.98  ...  \
0  100   4.08 -0.29   6.36   4.37  -2.38 -9.66  -0.73 -5.34   8.88  ...   
1   49   0.00  0.00   0.00   0.00   9.03  9.27   9.03  9.27   0.00  ...   
2   48   0.00  8.35   0.00   0.00   1.80  8.16  -2.82  6.21   0.00  ...   
3   91   8.50  4.61  -4.17  -5.39   1.36  1.60   7.04  4.61  -0.44  ...   
4  100  -6.17 -3.54   0.44  -8.50  -7.09 -4.32  -8.69 -0.87  -6.65  ...   

   2.82.2  99.18  99.19  99.20  99.21  99.22  -5.63  99.23  99.24  99.25  
0    2.82  -4.95  -0.29   7.86  -0.19  -2.14   3.06   0.34  -4.32   1.07  
1    0.00   0.00   0.00   9.08   0.00   0.00   0.00   0.00   0.00   0.00  
2    0.00   0.00   0.00   0.53   0.00   0.00   0.00   0.00   0.00   0.00  
3    5.19   5.58   4.27   5.19   5.73   1.55   3.11   6.55   1.80   1.60  
4   -3.54  -6.89  -0.68  -2.96  -2.18  -3.35   0.05  -9.08  -5.05  -3.45  

[5 rows x 101 columns]

In [100]:
##Transforming the data from wide to long 
df_02 = df_01.copy()
df_02.index = np.arange(1, len(df_01) + 1)
df_02

74  -7.82  8.79  -9.66  -8.16  -7.52  -8.5  -9.85  4.17  -8.98  ...  \
1      100   4.08 -0.29   6.36   4.37  -2.38 -9.66  -0.73 -5.34   8.88  ...   
2       49   0.00  0.00   0.00   0.00   9.03  9.27   9.03  9.27   0.00  ...   
3       48   0.00  8.35   0.00   0.00   1.80  8.16  -2.82  6.21   0.00  ...   
4       91   8.50  4.61  -4.17  -5.39   1.36  1.60   7.04  4.61  -0.44  ...   
5      100  -6.17 -3.54   0.44  -8.50  -7.09 -4.32  -8.69 -0.87  -6.65  ...   
...    ...    ...   ...    ...    ...    ...   ...    ...   ...    ...  ...   
24978  100   0.44  7.43   9.08   2.33   3.20  6.75  -8.79 -0.53  -8.74  ...   
24979   91   9.13 -8.16   8.59   9.08   0.87 -8.93  -3.50  5.78  -8.11  ...   
24980   39   0.00  0.00   0.00   0.00  -7.77  0.00   6.70 -6.75   0.00  ...   
24981   37   0.00  0.00   0.00   0.00  -9.71  0.00   4.56 -8.30   0.00  ...   
24982   72   2.43  2.67  -3.98   4.27  -2.28  7.33   2.33  4.56   6.75  ...   

       2.82.2  99.18  99.19  99.20  99.21  99.22  -5.63  99.23  99.24  99.25  
1        2.82  -4.95  -0.29   7.86  -0.19  -2.14   3.06   0.34  -4.32   1.07  
2        0.00   0.00   0.00   9.08   0.00   0.00   0.00   0.00   0.00   0.00  
3        0.00   0.00   0.00   0.53   0.00   0.00   0.00   0.00   0.00   0.00  
4        5.19   5.58   4.27   5.19   5.73   1.55   3.11   6.55   1.80   1.60  
5       -3.54  -6.89  -0.68  -2.96  -2.18  -3.35   0.05  -9.08  -5.05  -3.45  
...       ...    ...    ...    ...    ...    ...    ...    ...    ...    ...  
24978    8.83  -1.21   9.22  -6.70   8.45   9.03   6.55   8.69   8.79   7.43  
24979   -1.17  -5.73  -1.46   0.24   9.22  -8.20  -7.23  -8.59   9.13   8.45  
24980    0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00  
24981    0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00  
24982    0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00  

[24982 rows x 101 columns]

In [101]:
#Extract ratings columns
ratings = df_02.iloc[:, 1:]
ratings

-7.82  8.79  -9.66  -8.16  -7.52  -8.5  -9.85  4.17  -8.98  -4.76  ...  \
1       4.08 -0.29   6.36   4.37  -2.38 -9.66  -0.73 -5.34   8.88   9.22  ...   
2       0.00  0.00   0.00   0.00   9.03  9.27   9.03  9.27   0.00   0.00  ...   
3       0.00  8.35   0.00   0.00   1.80  8.16  -2.82  6.21   0.00   1.84  ...   
4       8.50  4.61  -4.17  -5.39   1.36  1.60   7.04  4.61  -0.44   5.73  ...   
5      -6.17 -3.54   0.44  -8.50  -7.09 -4.32  -8.69 -0.87  -6.65  -1.80  ...   
...      ...   ...    ...    ...    ...   ...    ...   ...    ...    ...  ...   
24978   0.44  7.43   9.08   2.33   3.20  6.75  -8.79 -0.53  -8.74   7.23  ...   
24979   9.13 -8.16   8.59   9.08   0.87 -8.93  -3.50  5.78  -8.11   4.90  ...   
24980   0.00  0.00   0.00   0.00  -7.77  0.00   6.70 -6.75   0.00   0.00  ...   
24981   0.00  0.00   0.00   0.00  -9.71  0.00   4.56 -8.30   0.00   0.00  ...   
24982   2.43  2.67  -3.98   4.27  -2.28  7.33   2.33  4.56   6.75   4.61  ...   

       2.82.2  99.18  99.19  99.20  99.21  99.22  -5.63  99.23  99.24  99.25  
1        2.82  -4.95  -0.29   7.86  -0.19  -2.14   3.06   0.34  -4.32   1.07  
2        0.00   0.00   0.00   9.08   0.00   0.00   0.00   0.00   0.00   0.00  
3        0.00   0.00   0.00   0.53   0.00   0.00   0.00   0.00   0.00   0.00  
4        5.19   5.58   4.27   5.19   5.73   1.55   3.11   6.55   1.80   1.60  
5       -3.54  -6.89  -0.68  -2.96  -2.18  -3.35   0.05  -9.08  -5.05  -3.45  
...       ...    ...    ...    ...    ...    ...    ...    ...    ...    ...  
24978    8.83  -1.21   9.22  -6.70   8.45   9.03   6.55   8.69   8.79   7.43  
24979   -1.17  -5.73  -1.46   0.24   9.22  -8.20  -7.23  -8.59   9.13   8.45  
24980    0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00  
24981    0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00  
24982    0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00  

[24982 rows x 100 columns]

In [102]:
df_03 = ratings.stack().reset_index()
df_03.columns=['UserId', 'jokeID', 'rating']

df_03

UserId jokeID  rating
0             1  -7.82    4.08
1             1   8.79   -0.29
2             1  -9.66    6.36
3             1  -8.16    4.37
4             1  -7.52   -2.38
...         ...    ...     ...
2498195   24982  99.22    0.00
2498196   24982  -5.63    0.00
2498197   24982  99.23    0.00
2498198   24982  99.24    0.00
2498199   24982  99.25    0.00

[2498200 rows x 3 columns]

In [103]:
##Sparsity of the matrix: 
##Sparsity Matrix = Number of Empty cells/Total number of cells

(df_03['rating']==0).sum() / 2498200

0.2774805860219358

There are about 27.75% of the cells that are empty

In [104]:
ratings.describe()

-7.82          8.79         -9.66         -8.16         -7.52  \
count  24982.000000  24982.000000  24982.000000  24982.000000  24982.000000   
mean       0.596024      0.151383      0.211276     -0.905036      0.403777   
std        4.218470      4.775975      4.409288      4.196370      5.254939   
min       -9.950000     -9.950000     -9.950000     -9.950000     -9.950000   
25%       -0.150000     -2.180000     -1.120000     -3.350000     -3.690000   
50%        0.000000      0.000000      0.000000      0.000000      1.120000   
75%        3.147500      3.250000      2.720000      0.000000      4.470000   
max        9.370000      9.370000      9.370000      9.370000      9.370000   

               -8.5         -9.85          4.17         -8.98         -4.76  \
count  24982.000000  24982.000000  24982.000000  24982.000000  24982.000000   
mean       1.304325     -0.429165     -0.624227     -0.350867      1.058801   
std        4.473837      5.494952      4.928955      4.107522      4.616374   
min       -9.950000     -9.950000     -9.950000     -9.950000     -9.950000   
25%       -0.390000     -5.290000     -4.610000     -1.750000     -0.830000   
50%        0.730000     -0.050000     -0.290000      0.000000      0.290000   
75%        4.660000      3.880000      3.060000      1.260000      4.470000   
max        9.470000      9.610000      9.760000      9.900000      9.370000   

       ...        2.82.2         99.18         99.19         99.20  \
count  ...  24982.000000  24982.000000  24982.000000  24982.000000   
mean   ...      0.789121      0.466996      1.001135      0.479126   
std    ...      3.301463      3.276987      3.163203      3.334715   
min    ...     -9.950000     -9.950000     -9.950000     -9.950000   
25%    ...      0.000000      0.000000      0.000000      0.000000   
50%    ...      0.000000      0.000000      0.000000      0.000000   
75%    ...      0.780000      0.000000      1.750000      0.150000   
max    ...      9.560000      9.420000      9.370000      9.420000   

              99.21         99.22         -5.63         99.23         99.24  \
count  24982.000000  24982.000000  24982.000000  24982.000000  24982.000000   
mean       0.426974      0.634272      0.705047      0.327764     -0.013645   
std        3.351267      3.215406      3.363661      3.544700      3.421007   
min       -9.950000     -9.950000     -9.950000     -9.950000     -9.950000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.100000      0.870000      1.120000      0.000000      0.000000   
max        9.370000      9.370000      9.370000      9.370000      9.370000   

              99.25  
count  24982.000000  
mean       0.539184  
std        3.408928  
min       -9.950000  
25%        0.000000  
50%        0.000000  
75%        0.100000  
max        9.370000  

[8 rows x 100 columns]

In [105]:
#Fit StandardScaler into ratings
from sklearn.preprocessing import StandardScaler

df_03[['rating']] = StandardScaler().fit_transform(df_03[['rating']])
df_03

UserId jokeID    rating
0             1  -7.82  0.769346
1             1   8.79 -0.207329
2             1  -9.66  1.278915
3             1  -8.16  0.834159
4             1  -7.52 -0.674435
...         ...    ...       ...
2498195   24982  99.22 -0.142516
2498196   24982  -5.63 -0.142516
2498197   24982  99.23 -0.142516
2498198   24982  99.24 -0.142516
2498199   24982  99.25 -0.142516

[2498200 rows x 3 columns]

### MovieLens

### Survery Dataset

In [109]:
#Loading the Data from Github Repository
username = "harisx91"
path = "Recommender_System_DAV_6300/master/Module_1/M1_Rating.csv"
df_survey = pd.read_csv(f'https://raw.githubusercontent.com/{username}/{path}')
df_survey

UserID  The Whale  Top Gun  I Came By  The Godfather  Top Gun: Maverick  \
0       1        1.0      4.0        5.0            NaN                  4   
1       2        NaN      NaN        4.0            4.0                  5   
2       3        3.0      5.0        2.0            NaN                  5   
3       4        NaN      NaN        NaN            5.0                  3   
4       5        2.0      4.0        3.0            3.0                  5   
5       6        3.0      4.0        NaN            3.0                  4   
6       7        NaN      2.0        NaN            NaN                  2   

   Spider Man  
0         3.0  
1         5.0  
2         4.0  
3         4.0  
4         3.0  
5         5.0  
6         NaN

In [110]:
df_survey_01 = df_survey.fillna(0)
df_survey_01

UserID  The Whale  Top Gun  I Came By  The Godfather  Top Gun: Maverick  \
0       1        1.0      4.0        5.0            0.0                  4   
1       2        0.0      0.0        4.0            4.0                  5   
2       3        3.0      5.0        2.0            0.0                  5   
3       4        0.0      0.0        0.0            5.0                  3   
4       5        2.0      4.0        3.0            3.0                  5   
5       6        3.0      4.0        0.0            3.0                  4   
6       7        0.0      2.0        0.0            0.0                  2   

   Spider Man  
0         3.0  
1         5.0  
2         4.0  
3         4.0  
4         3.0  
5         5.0  
6         0.0

In [112]:
df_survey_02 = df_survey_01.copy()
df_survey_02.index = np.arange(1, len(df_survey_01) +1)
df_survey_02

UserID  The Whale  Top Gun  I Came By  The Godfather  Top Gun: Maverick  \
1       1        1.0      4.0        5.0            0.0                  4   
2       2        0.0      0.0        4.0            4.0                  5   
3       3        3.0      5.0        2.0            0.0                  5   
4       4        0.0      0.0        0.0            5.0                  3   
5       5        2.0      4.0        3.0            3.0                  5   
6       6        3.0      4.0        0.0            3.0                  4   
7       7        0.0      2.0        0.0            0.0                  2   

   Spider Man  
1         3.0  
2         5.0  
3         4.0  
4         4.0  
5         3.0  
6         5.0  
7         0.0

In [113]:
df_survey_03 = df_survey_02.drop(['UserID'], axis=1)

df_survey = df_survey_03.stack().reset_index()
df_survey.columns=['UserId', 'movie', 'rating']

df_survey

UserId              movie  rating
0        1          The Whale     1.0
1        1            Top Gun     4.0
2        1          I Came By     5.0
3        1      The Godfather     0.0
4        1  Top Gun: Maverick     4.0
5        1         Spider Man     3.0
6        2          The Whale     0.0
7        2            Top Gun     0.0
8        2          I Came By     4.0
9        2      The Godfather     4.0
10       2  Top Gun: Maverick     5.0
11       2         Spider Man     5.0
12       3          The Whale     3.0
13       3            Top Gun     5.0
14       3          I Came By     2.0
15       3      The Godfather     0.0
16       3  Top Gun: Maverick     5.0
17       3         Spider Man     4.0
18       4          The Whale     0.0
19       4            Top Gun     0.0
20       4          I Came By     0.0
21       4      The Godfather     5.0
22       4  Top Gun: Maverick     3.0
23       4         Spider Man     4.0
24       5          The Whale     2.0
25       5            Top Gun     4.0
26       5          I Came By     3.0
27       5      The Godfather     3.0
28       5  Top Gun: Maverick     5.0
29       5         Spider Man     3.0
30       6          The Whale     3.0
31       6            Top Gun     4.0
32       6          I Came By     0.0
33       6      The Godfather     3.0
34       6  Top Gun: Maverick     4.0
35       6         Spider Man     5.0
36       7          The Whale     0.0
37       7            Top Gun     2.0
38       7          I Came By     0.0
39       7      The Godfather     0.0
40       7  Top Gun: Maverick     2.0
41       7         Spider Man     0.0

Sparsity of the matrix

In [115]:
12/42

0.2857142857142857

28.57% of the survey dataset is sparse